jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -651,83 +684,83 @@ public String toString()
-                                job = _jobs.take();
-                            else
+                            // maybe we should shrink?
+                            int size = _threadsStarted.get();
+                            if (size > _minThreads)
                             {
-                                // maybe we should shrink?
-                                final int size = _threadsStarted.get();
-                                if (size > _minThreads)
+                                long last = _lastShrink.get();
+                                long now = System.nanoTime();
+                                if (last == 0 || (now - last) > TimeUnit.MILLISECONDS.toNanos(_idleTimeout))
                                 {
-                                    long last = _lastShrink.get();
-                                    long now = System.nanoTime();
-                                    if (last == 0 || (now - last) > TimeUnit.MILLISECONDS.toNanos(_idleTimeout))
+                                    if (_lastShrink.compareAndSet(last, now))
                                     {
-                                        if (_lastShrink.compareAndSet(last, now) && _threadsStarted.compareAndSet(size, size - 1))
-                                        {
-                                            shrink = true;
-                                            break loop;
-                                        }
+                                        _threadsStarted.decrementAndGet();
+                                        shrink = true;
+                                        break loop;
                                     }
                                 }
-                                job = idleJobPoll();
                             }
+
+                            job = _jobs.poll(_idleTimeout, TimeUnit.MILLISECONDS);
                         }
                     }
-                    finally
+
+                    // run job
+                    if (job != null)
                     {
-                        if (_threadsIdle.decrementAndGet() == 0)
+                        if (idle)
                         {
-                            startThreads(1);
+                            idle = false;
+                            if (_threadsIdle.decrementAndGet() == 0)
+                                startThreads(1);
+                        }
+
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("run {}", job);
+                        runJob(job);
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("ran {}", job);
+                        if (Thread.interrupted())
+                        {
+                            ignore = true;
+                            break loop;
                         }
                     }
+
+                    if (!isRunning())
+                        break loop;
```

Comment:
We break the loop here, but because we are not running, we won't decrement `_threadsStarted`.
So this is one more reason to move the decrement of `_threadsStarted` in the finally block and doing it unconditionally, as the above comment also suggests.